File: selenium-tests/data_grid_test_suite.py
"""
Comprehensive Selenium Test Suite for Angular Data Grid Application

This test suite validates all UI components and functionality:
- Data grid rendering and pagination
- Row selection with checkboxes
- Export functionality
- Comments and attachments
- Row-level actions dropdown
- Scroll-based lazy loading
"""

import time
import unittest
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.chrome.options import Options
from selenium.common.exceptions import TimeoutException, NoSuchElementException

class DataGridTestSuite(unittest.TestCase):
    """Test suite for Angular Data Grid Application"""
    
    @classmethod
    def setUpClass(cls):
        """Set up the test environment"""
        print("🚀 Setting up Selenium Test Suite...")
        
        # Chrome driver setup
        chrome_options = Options()
        chrome_options.add_argument("--start-maximized")
        chrome_options.add_argument("--disable-notifications")
        chrome_options.add_argument("--disable-popup-blocking")
        chrome_options.add_argument("--disable-web-security")
        chrome_options.add_argument("--allow-running-insecure-content")
        
        cls.driver = webdriver.Chrome(options=chrome_options)
        cls.wait = WebDriverWait(cls.driver, 10)
        cls.actions = ActionChains(cls.driver)
        cls.base_url = "http://localhost:4200"
        
        # Navigate to application
        cls.driver.get(cls.base_url)
        cls.wait_for_page_load()
        
        print("✅ Test environment ready!")
    
    @classmethod
    def tearDownClass(cls):
        """Clean up after tests"""
        if cls.driver:
            cls.driver.quit()
        print("🏁 Test suite completed!")
    
    def wait_for_page_load(self):
        """Wait for page to fully load"""
        self.wait.until(lambda driver: driver.execute_script("return document.readyState") == "complete")
    
    def test_01_application_loads(self):
        """Test 1: Verify Application Loads Successfully"""
        print("\n=== Test 1: Application Load ===")
        
        # Verify page title
        title = self.driver.title
        self.assertEqual(title, "Data Grid", "Page title should be 'Data Grid'")
        
        # Verify main container is present
        container = self.wait.until(EC.presence_of_element_located((By.CLASS_NAME, "data-grid-container")))
        self.assertTrue(container.is_displayed(), "Data grid container should be visible")
        
        print("✅ Application loaded successfully")
    
    def test_02_data_grid_structure(self):
        """Test 2: Verify Data Grid Structure and Headers"""
        print("\n=== Test 2: Data Grid Structure ===")
        
        # Verify table is present
        table = self.wait.until(EC.presence_of_element_located((By.CLASS_NAME, "data-table")))
        self.assertTrue(table.is_displayed(), "Data table should be visible")
        
        # Verify all column headers
        expected_headers = [
            "Select", "Name", "Email", "Department", "Position", 
            "Salary", "Start Date", "Status", "Comments & Attachments", "Actions"
        ]
        
        headers = self.driver.find_elements(By.TAG_NAME, "th")
        self.assertEqual(len(headers), len(expected_headers), 
                        f"Should have {len(expected_headers)} column headers")
        
        for i, expected_header in enumerate(expected_headers):
            self.assertEqual(headers[i].text, expected_header,
                           f"Header {i} should be '{expected_header}'")
        
        print("✅ Data grid structure validated")
    
    def test_03_initial_data_loading(self):
        """Test 3: Verify Initial Data Loading"""
        print("\n=== Test 3: Initial Data Loading ===")
        
        # Wait for data rows to load
        rows = self.wait.until(EC.presence_of_all_elements_located((By.CSS_SELECTOR, "tbody tr.data-row")))
        
        self.assertGreater(len(rows), 0, "Should have at least one data row")
        self.assertLessEqual(len(rows), 20, "Initial load should not exceed 20 rows")
        
        # Verify first row has all expected data
        first_row = rows[0]
        cells = first_row.find_elements(By.TAG_NAME, "td")
        self.assertEqual(len(cells), 10, "Each row should have 10 cells")
        
        # Verify checkbox is present
        checkbox = first_row.find_element(By.CSS_SELECTOR, "input[type='checkbox']")
        self.assertTrue(checkbox.is_displayed(), "Checkbox should be visible")
        
        print(f"✅ Initial data loading validated - {len(rows)} rows loaded")
    
    def test_04_checkbox_selection(self):
        """Test 4: Test Checkbox Selection and Deselection"""
        print("\n=== Test 4: Checkbox Selection ===")
        
        rows = self.driver.find_elements(By.CSS_SELECTOR, "tbody tr.data-row")
        self.assertGreaterEqual(len(rows), 3, "Need at least 3 rows for testing")
        
        # Select first row
        first_checkbox = rows[0].find_element(By.CSS_SELECTOR, "input[type='checkbox']")
        first_checkbox.click()
        
        # Verify selected IDs display appears
        selected_ids_display = self.wait.until(EC.presence_of_element_located((By.CLASS_NAME, "selected-ids-display")))
        self.assertTrue(selected_ids_display.is_displayed(), "Selected IDs should be displayed")
        
        # Select second row
        second_checkbox = rows[1].find_element(By.CSS_SELECTOR, "input[type='checkbox']")
        second_checkbox.click()
        
        # Verify both IDs are shown
        selected_text = selected_ids_display.text
        self.assertIn("Selected Row IDs:", selected_text, "Should show selected row IDs")
        
        # Deselect first row
        first_checkbox.click()
        
        # Verify only second row ID remains
        selected_text = selected_ids_display.text
        self.assertNotIn("1,", selected_text, "First row ID should be removed")
        
        print("✅ Checkbox selection and deselection validated")
    
    def test_05_export_functionality(self):
        """Test 5: Test Export Functionality"""
        print("\n=== Test 5: Export Functionality ===")
        
        # Find export buttons
        export_csv_btn = self.wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR, ".export-btn[title='Export to CSV']")))
        export_txt_btn = self.driver.find_element(By.CSS_SELECTOR, ".export-btn[title='Export to Text']")
        
        self.assertTrue(export_csv_btn.is_displayed(), "CSV export button should be visible")
        self.assertTrue(export_txt_btn.is_displayed(), "Text export button should be visible")
        
        # Test CSV export
        export_csv_btn.click()
        
        print("✅ Export buttons are functional")
    
    def test_06_comments_modal(self):
        """Test 6: Test Comments and Attachments Modal"""
        print("\n=== Test 6: Comments Modal ===")
        
        # Find first "Add Comment" button
        add_comment_btn = self.wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR, ".add-comment-btn")))
        add_comment_btn.click()
        
        # Verify modal opens
        modal = self.wait.until(EC.presence_of_element_located((By.CLASS_NAME, "modal-content")))
        self.assertTrue(modal.is_displayed(), "Comment modal should be visible")
        
        # Verify modal elements
        textarea = modal.find_element(By.CSS_SELECTOR, "textarea")
        file_input = modal.find_element(By.CSS_SELECTOR, "input[type='file']")
        submit_btn = modal.find_element(By.CSS_SELECTOR, ".btn-submit")
        cancel_btn = modal.find_element(By.CSS_SELECTOR, ".btn-cancel")
        
        self.assertTrue(textarea.is_displayed(), "Comment textarea should be visible")
        self.assertTrue(file_input.is_displayed(), "File input should be visible")
        self.assertTrue(submit_btn.is_displayed(), "Submit button should be visible")
        self.assertTrue(cancel_btn.is_displayed(), "Cancel button should be visible")
        
        # Test adding a comment
        textarea.send_keys("This is a test comment from Selenium automation")
        
        # Submit button should be enabled
        self.assertTrue(submit_btn.is_enabled(), "Submit button should be enabled with text")
        
        # Close modal
        cancel_btn.click()
        
        # Verify modal closes
        self.wait.until(EC.invisibility_of_element_located((By.CLASS_NAME, "modal-content")))
        
        print("✅ Comments modal functionality validated")
    
    def test_07_row_actions_dropdown(self):
        """Test 7: Test Row Actions Dropdown"""
        print("\n=== Test 7: Row Actions Dropdown ===")
        
        # Find first ellipsis button
        ellipsis_btn = self.wait.until(EC.element_to_be_clickable((By.CSS_SELECTOR, ".ellipsis-btn")))
        ellipsis_btn.click()
        
        # Verify dropdown appears
        dropdown = self.wait.until(EC.presence_of_element_located((By.CLASS_NAME, "dropdown-menu")))
        self.assertTrue(dropdown.is_displayed(), "Dropdown menu should be visible")
        
        # Verify dropdown items
        dropdown_items = dropdown.find_elements(By.CLASS_NAME, "dropdown-item")
        self.assertEqual(len(dropdown_items), 3, "Should have 3 dropdown items")
        
        expected_actions = ["Action A", "Action B", "Action C"]
        for i, expected_action in enumerate(expected_actions):
            self.assertEqual(dropdown_items[i].text.strip(), expected_action,
                           f"Dropdown item {i} should be '{expected_action}'")
        
        # Test clicking Action A
        dropdown_items[0].click()
        
        # Handle alert if it appears
        try:
            alert = self.driver.switch_to.alert
            alert.accept()
        except:
            pass  # Alert might not appear in headless mode
        
        # Verify dropdown closes
        self.wait.until(EC.invisibility_of_element_located((By.CLASS_NAME, "dropdown-menu")))
        
        print("✅ Row actions dropdown validated")
    
    def test_08_scroll_pagination(self):
        """Test 8: Test Scroll-based Pagination (Lazy Loading)"""
        print("\n=== Test 8: Scroll Pagination ===")
        
        # Get initial row count
        initial_rows = self.driver.find_elements(By.CSS_SELECTOR, "tbody tr.data-row")
        initial_count = len(initial_rows)
        
        # Scroll to bottom to trigger lazy loading
        self.driver.execute_script("window.scrollTo(0, document.body.scrollHeight);")
        
        # Wait for loading indicator (might be too fast to catch)
        try:
            loading_indicator = self.wait.until(EC.presence_of_element_located((By.CLASS_NAME, "loading-indicator")))
            self.assertTrue(loading_indicator.is_displayed(), "Loading indicator should appear")
        except TimeoutException:
            pass  # Loading might be too fast to catch
        
        # Wait for new rows to load
        self.wait.until(lambda driver: len(driver.find_elements(By.CSS_SELECTOR, "tbody tr.data-row")) > initial_count)
        
        # Verify more rows are loaded
        new_rows = self.driver.find_elements(By.CSS_SELECTOR, "tbody tr.data-row")
        self.assertGreater(len(new_rows), initial_count, "More rows should be loaded after scroll")
        
        print(f"✅ Scroll pagination validated - {len(new_rows)} total rows")
    
    def test_09_responsive_design(self):
        """Test 9: Test Responsive Design"""
        print("\n=== Test 9: Responsive Design ===")
        
        # Test desktop viewport
        self.driver.set_window_size(1200, 800)
        time.sleep(1)
        
        table = self.driver.find_element(By.CLASS_NAME, "data-table")
        self.assertTrue(table.is_displayed(), "Table should be visible on desktop")
        
        # Test mobile viewport
        self.driver.set_window_size(375, 667)
        time.sleep(1)
        
        table = self.driver.find_element(By.CLASS_NAME, "data-table")
        self.assertTrue(table.is_displayed(), "Table should be visible on mobile")
        
        # Reset to full screen
        self.driver.maximize_window()
        
        print("✅ Responsive design validated")
    
    def test_10_data_integrity(self):
        """Test 10: Test Data Integrity"""
        print("\n=== Test 10: Data Integrity ===")
        
        rows = self.driver.find_elements(By.CSS_SELECTOR, "tbody tr.data-row")
        
        for i in range(min(5, len(rows))):
            row = rows[i]
            cells = row.find_elements(By.TAG_NAME, "td")
            
            # Verify each row has 10 cells
            self.assertEqual(len(cells), 10, f"Row {i} should have 10 cells")
            
            # Verify name is not empty
            name = cells[1].text
            self.assertNotEqual(name, "", f"Name should not be empty in row {i}")
            
            # Verify email format
            email = cells[2].text
            self.assertIn("@", email, f"Email should contain @ in row {i}")
            
            # Verify salary is numeric
            salary = cells[5].text
            self.assertTrue(salary.startswith("$"), f"Salary should start with $ in row {i}")
        
        print("✅ Data integrity validated")

if __name__ == "__main__":
    # Run the test suite
    unittest.main(verbosity=2)

File: selenium-tests/DataGridTestSuite.java
package com.datagrid.tests;

import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.chrome.ChromeDriver;
import org.openqa.selenium.chrome.ChromeOptions;
import org.openqa.selenium.support.ui.WebDriverWait;
import org.openqa.selenium.support.ui.ExpectedConditions;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.JavascriptExecutor;
import org.testng.annotations.*;
import org.testng.Assert;
import java.time.Duration;
import java.util.List;
import java.util.ArrayList;

/**
 * Comprehensive Selenium Test Suite for Angular Data Grid Application
 * 
 * This test suite validates all UI components and functionality:
 * - Data grid rendering and pagination
 * - Row selection with checkboxes
 * - Export functionality
 * - Comments and attachments
 * - Row-level actions dropdown
 * - Scroll-based lazy loading
 */
public class DataGridTestSuite {
    
    private WebDriver driver;
    private WebDriverWait wait;
    private Actions actions;
    private JavascriptExecutor js;
    private static final String BASE_URL = "D";
    private static final int TIMEOUT = 10;
    
    @BeforeClass
    public void setUp() {
        // Chrome driver setup
        ChromeOptions options = new ChromeOptions();
        options.addArguments("--start-maximized");
        options.addArguments("--disable-notifications");
        options.addArguments("--disable-popup-blocking");
        
        driver = new ChromeDriver(options);
        wait = new WebDriverWait(driver, Duration.ofSeconds(TIMEOUT));
        actions = new Actions(driver);
        js = (JavascriptExecutor) driver;
        
        // Navigate to application
        driver.get(BASE_URL);
        waitForPageLoad();
    }
    
    @AfterClass
    public void tearDown() {
        if (driver != null) {
            driver.quit();
        }
    }
    
    /**
     * Test 1: Verify Application Loads Successfully
     */
    @Test(priority = 1)
    public void testApplicationLoads() {
        System.out.println("=== Test 1: Application Load ===");
        
        // Verify page title
        String title = driver.getTitle();
        Assert.assertEquals(title, "Data Grid", "Page title should be 'Data Grid'");
        
        // Verify main container is present
        WebElement container = wait.until(ExpectedConditions.presenceOfElementLocated(
            By.className("data-grid-container")));
        Assert.assertTrue(container.isDisplayed(), "Data grid container should be visible");
        
        System.out.println("✅ Application loaded successfully");
    }
    
    /**
     * Test 2: Verify Data Grid Structure and Headers
     */
    @Test(priority = 2)
    public void testDataGridStructure() {
        System.out.println("=== Test 2: Data Grid Structure ===");
        
        // Verify table is present
        WebElement table = wait.until(ExpectedConditions.presenceOfElementLocated(
            By.className("data-table")));
        Assert.assertTrue(table.isDisplayed(), "Data table should be visible");
        
        // Verify all column headers
        String[] expectedHeaders = {
            "Select", "Name", "Email", "Department", "Position", 
            "Salary", "Start Date", "Status", "Comments & Attachments", "Actions"
        };
        
        List<WebElement> headers = driver.findElements(By.tagName("th"));
        Assert.assertEquals(headers.size(), expectedHeaders.length, 
            "Should have " + expectedHeaders.length + " column headers");
        
        for (int i = 0; i < expectedHeaders.length; i++) {
            Assert.assertEquals(headers.get(i).getText(), expectedHeaders[i],
                "Header " + i + " should be '" + expectedHeaders[i] + "'");
        }
        
        System.out.println("✅ Data grid structure validated");
    }
    
    /**
     * Test 3: Verify Initial Data Loading
     */
    @Test(priority = 3)
    public void testInitialDataLoading() {
        System.out.println("=== Test 3: Initial Data Loading ===");
        
        // Wait for data rows to load
        List<WebElement> rows = wait.until(ExpectedConditions.presenceOfAllElementsLocatedBy(
            By.cssSelector("tbody tr.data-row")));
        
        Assert.assertTrue(rows.size() > 0, "Should have at least one data row");
        Assert.assertTrue(rows.size() <= 20, "Initial load should not exceed 20 rows");
        
        // Verify first row has all expected data
        WebElement firstRow = rows.get(0);
        List<WebElement> cells = firstRow.findElements(By.tagName("td"));
        Assert.assertEquals(cells.size(), 10, "Each row should have 10 cells");
        
        // Verify checkbox is present
        WebElement checkbox = firstRow.findElement(By.cssSelector("input[type='checkbox']"));
        Assert.assertTrue(checkbox.isDisplayed(), "Checkbox should be visible");
        
        System.out.println("✅ Initial data loading validated - " + rows.size() + " rows loaded");
    }
    
    /**
     * Test 4: Test Checkbox Selection and Deselection
     */
    @Test(priority = 4)
    public void testCheckboxSelection() {
        System.out.println("=== Test 4: Checkbox Selection ===");
        
        List<WebElement> rows = driver.findElements(By.cssSelector("tbody tr.data-row"));
        Assert.assertTrue(rows.size() >= 3, "Need at least 3 rows for testing");
        
        // Select first row
        WebElement firstCheckbox = rows.get(0).findElement(By.cssSelector("input[type='checkbox']"));
        firstCheckbox.click();
        
        // Verify selected IDs display appears
        WebElement selectedIdsDisplay = wait.until(ExpectedConditions.presenceOfElementLocated(
            By.className("selected-ids-display")));
        Assert.assertTrue(selectedIdsDisplay.isDisplayed(), "Selected IDs should be displayed");
        
        // Select second row
        WebElement secondCheckbox = rows.get(1).findElement(By.cssSelector("input[type='checkbox']"));
        secondCheckbox.click();
        
        // Verify both IDs are shown
        String selectedText = selectedIdsDisplay.getText();
        Assert.assertTrue(selectedText.contains("Selected Row IDs:"), "Should show selected row IDs");
        
        // Deselect first row
        firstCheckbox.click();
        
        // Verify only second row ID remains
        selectedText = selectedIdsDisplay.getText();
        Assert.assertFalse(selectedText.contains("1,"), "First row ID should be removed");
        
        System.out.println("✅ Checkbox selection and deselection validated");
    }
    
    /**
     * Test 5: Test Export Functionality
     */
    @Test(priority = 5)
    public void testExportFunctionality() {
        System.out.println("=== Test 5: Export Functionality ===");
        
        // Find export buttons
        WebElement exportCsvBtn = wait.until(ExpectedConditions.elementToBeClickable(
            By.cssSelector(".export-btn[title='Export to CSV']")));
        WebElement exportTxtBtn = driver.findElement(By.cssSelector(".export-btn[title='Export to Text']"));
        
        Assert.assertTrue(exportCsvBtn.isDisplayed(), "CSV export button should be visible");
        Assert.assertTrue(exportTxtBtn.isDisplayed(), "Text export button should be visible");
        
        // Test CSV export
        exportCsvBtn.click();
        
        // Verify download started (this would need additional setup for actual file download testing)
        System.out.println("✅ Export buttons are functional");
    }
    
    /**
     * Test 6: Test Comments and Attachments Modal
     */
    @Test(priority = 6)
    public void testCommentsModal() {
        System.out.println("=== Test 6: Comments Modal ===");
        
        // Find first "Add Comment" button
        WebElement addCommentBtn = wait.until(ExpectedConditions.elementToBeClickable(
            By.cssSelector(".add-comment-btn")));
        addCommentBtn.click();
        
        // Verify modal opens
        WebElement modal = wait.until(ExpectedConditions.presenceOfElementLocated(
            By.className("modal-content")));
        Assert.assertTrue(modal.isDisplayed(), "Comment modal should be visible");
        
        // Verify modal elements
        WebElement textarea = modal.findElement(By.cssSelector("textarea"));
        WebElement fileInput = modal.findElement(By.cssSelector("input[type='file']"));
        WebElement submitBtn = modal.findElement(By.cssSelector(".btn-submit"));
        WebElement cancelBtn = modal.findElement(By.cssSelector(".btn-cancel"));
        
        Assert.assertTrue(textarea.isDisplayed(), "Comment textarea should be visible");
        Assert.assertTrue(fileInput.isDisplayed(), "File input should be visible");
        Assert.assertTrue(submitBtn.isDisplayed(), "Submit button should be visible");
        Assert.assertTrue(cancelBtn.isDisplayed(), "Cancel button should be visible");
        
        // Test adding a comment
        textarea.sendKeys("This is a test comment from Selenium automation");
        
        // Submit button should be enabled
        Assert.assertTrue(submitBtn.isEnabled(), "Submit button should be enabled with text");
        
        // Close modal
        cancelBtn.click();
        
        // Verify modal closes
        wait.until(ExpectedConditions.invisibilityOf(modal));
        
        System.out.println("✅ Comments modal functionality validated");
    }
    
    /**
     * Test 7: Test Row Actions Dropdown
     */
    @Test(priority = 7)
    public void testRowActionsDropdown() {
        System.out.println("=== Test 7: Row Actions Dropdown ===");
        
        // Find first ellipsis button
        WebElement ellipsisBtn = wait.until(ExpectedConditions.elementToBeClickable(
            By.cssSelector(".ellipsis-btn")));
        ellipsisBtn.click();
        
        // Verify dropdown appears
        WebElement dropdown = wait.until(ExpectedConditions.presenceOfElementLocated(
            By.className("dropdown-menu")));
        Assert.assertTrue(dropdown.isDisplayed(), "Dropdown menu should be visible");
        
        // Verify dropdown items
        List<WebElement> dropdownItems = dropdown.findElements(By.className("dropdown-item"));
        Assert.assertEquals(dropdownItems.size(), 3, "Should have 3 dropdown items");
        
        String[] expectedActions = {"Action A", "Action B", "Action C"};
        for (int i = 0; i < dropdownItems.size(); i++) {
            Assert.assertEquals(dropdownItems.get(i).getText().trim(), expectedActions[i],
                "Dropdown item " + i + " should be '" + expectedActions[i] + "'");
        }
        
        // Test clicking Action A
        dropdownItems.get(0).click();
        
        // Verify alert appears (you might need to handle alert)
        try {
            driver.switchTo().alert().accept();
        } catch (Exception e) {
            // Alert might not appear in headless mode
        }
        
        // Verify dropdown closes
        wait.until(ExpectedConditions.invisibilityOf(dropdown));
        
        System.out.println("✅ Row actions dropdown validated");
    }
    
    /**
     * Test 8: Test Scroll-based Pagination (Lazy Loading)
     */
    @Test(priority = 8)
    public void testScrollPagination() {
        System.out.println("=== Test 8: Scroll Pagination ===");
        
        // Get initial row count
        List<WebElement> initialRows = driver.findElements(By.cssSelector("tbody tr.data-row"));
        int initialCount = initialRows.size();
        
        // Scroll to bottom to trigger lazy loading
        js.executeScript("window.scrollTo(0, document.body.scrollHeight);");
        
        // Wait for loading indicator
        try {
            WebElement loadingIndicator = wait.until(ExpectedConditions.presenceOfElementLocated(
                By.className("loading-indicator")));
            Assert.assertTrue(loadingIndicator.isDisplayed(), "Loading indicator should appear");
        } catch (Exception e) {
            // Loading might be too fast to catch
        }
        
        // Wait for new rows to load
        wait.until(ExpectedConditions.numberOfElementsToBeMoreThan(
            By.cssSelector("tbody tr.data-row"), initialCount));
        
        // Verify more rows are loaded
        List<WebElement> newRows = driver.findElements(By.cssSelector("tbody tr.data-row"));
        Assert.assertTrue(newRows.size() > initialCount, "More rows should be loaded after scroll");
        
        System.out.println("✅ Scroll pagination validated - " + newRows.size() + " total rows");
    }
    
    /**
     * Test 9: Test Responsive Design
     */
    @Test(priority = 9)
    public void testResponsiveDesign() {
        System.out.println("=== Test 9: Responsive Design ===");
        
        // Test different viewport sizes
        driver.manage().window().setSize(new org.openqa.selenium.Dimension(1200, 800));
        Thread.sleep(1000);
        
        WebElement table = driver.findElement(By.className("data-table"));
        Assert.assertTrue(table.isDisplayed(), "Table should be visible on desktop");
        
        // Test mobile viewport
        driver.manage().window().setSize(new org.openqa.selenium.Dimension(375, 667));
        Thread.sleep(1000);
        
        table = driver.findElement(By.className("data-table"));
        Assert.assertTrue(table.isDisplayed(), "Table should be visible on mobile");
        
        // Reset to full screen
        driver.manage().window().maximize();
        
        System.out.println("✅ Responsive design validated");
    }
    
    /**
     * Test 10: Test Data Integrity
     */
    @Test(priority = 10)
    public void testDataIntegrity() {
        System.out.println("=== Test 10: Data Integrity ===");
        
        List<WebElement> rows = driver.findElements(By.cssSelector("tbody tr.data-row"));
        
        for (int i = 0; i < Math.min(5, rows.size()); i++) {
            WebElement row = rows.get(i);
            List<WebElement> cells = row.findElements(By.tagName("td"));
            
            // Verify each row has 10 cells
            Assert.assertEquals(cells.size(), 10, "Row " + i + " should have 10 cells");
            
            // Verify name is not empty
            String name = cells.get(1).getText();
            Assert.assertFalse(name.isEmpty(), "Name should not be empty in row " + i);
            
            // Verify email format
            String email = cells.get(2).getText();
            Assert.assertTrue(email.contains("@"), "Email should contain @ in row " + i);
            
            // Verify salary is numeric
            String salary = cells.get(5).getText();
            Assert.assertTrue(salary.startsWith("$"), "Salary should start with $ in row " + i);
        }
        
        System.out.println("✅ Data integrity validated");
    }
    
    /**
     * Helper method to wait for page load
     */
    private void waitForPageLoad() {
        wait.until(webDriver -> 
            js.executeScript("return document.readyState").equals("complete"));
    }
    
    /**
     * Helper method to scroll to element
     */
    private void scrollToElement(WebElement element) {
        js.executeScript("arguments[0].scrollIntoView(true);", element);
    }
    
    /**
     * Helper method to take screenshot
     */
    private void takeScreenshot(String testName) {
        try {
            // Screenshot functionality would go here
            System.out.println("📸 Screenshot taken for: " + testName);
        } catch (Exception e) {
            System.out.println("❌ Failed to take screenshot: " + e.getMessage());
        }
    }
}
